### ISSUE TITLE
Nintendo Switch UI textures don't seem to be unswizzled properly

### ISSUE THREAD

**Issue by CosmicDreamsOfCode** [gildor2/UModel#122]
Happens in both Yoshi's Crafted World and Octopath Traveller.
![image](https://user-images.githubusercontent.com/33284629/52921695-c8553a00-3311-11e9-8fe7-f2fdf09f0a9b.png)

I should mention too, not an issue but Yoshi's Crafted World does indeed work.
Seems to be 4.18, with everything working.

**Comment by gildor2** [comment]
Yes, I'm aware about the problems with non-power-of-two UI textures for NSW. I even mentioned that in forum thread:
http://www.gildor.org/smf/index.php/topic,6257.msg31496.html#msg31496

Previously I spent time trying to fix that, but wasn't able to find a solution. Documentation and sample NSW decoders are quiet about this situation. Actually umodel already contains some empiric code to make MOST things working, but I didn't get any luck with NP2 textures.

**Comment by masagrator** [comment]
umodel viewer is showing different results than converted file.
Converted doesn't have glithes, but still is wrongly tiled.

![image](https://user-images.githubusercontent.com/8557858/55739803-16411100-5a2a-11e9-8ab0-99878e387858.png)

**Comment by gildor2** [comment]
What do you mean with "different results"? Different background? If yes, then I'd say that if you'll open this TGA say in Paint.NET you'd see a checkerboard background. Every little application displays transparency in its own way. On screenshot from UModel you just see color pixels, I'm not enabling transparency when rendering. I have special mode for texture rendering which is activated with key "C".

**Comment by gildor2** [comment]
(accidentally pressed "close and comment" :))

**Comment by masagrator** [comment]
You don't understand. I said that what is in first post is showing something different to exported file. So glitches are because of transparency.

**Comment by masagrator** [comment]
And tiling is the result of wrong texture size. Switch textures are rounding pixels to 128, 256, 512, 1024, 2048 in width and height, but it contains info about real resolution. So if you will round size of texture to the upper ones, you will get proper texture.

**Comment by masagrator** [comment]
And this tiling is the result of using different size for texture and different original resolution.
F.e. this MC uses resolution 622x720, but this is what we get if we use 650x800
![UiTX_Title_CharaSelect_Chara_Ulr](https://user-images.githubusercontent.com/8557858/55744811-1941fe80-5a36-11e9-90cd-b4a7bfa5593e.PNG)

**Comment by gildor2** [comment]
That would be nice if you could contribute a pull request for that. I have too many things to do, and UModel has lowest priority. And it has its own task list, where NSW texture fix is rather in "backlog" because that'd be require me to spent too much time doing trial-and-error. I already spent a day or two I think doing what it already can do - seeing that texture works, then seeing that mips are not. Then doing a fix, walking over all game's textures and seeing other problems. And - there's no relevant documentation about that. All I have found is "official Tegra documentation" which is too sparse, and a few NEW texture decoders - which assumes that the file has full header with full information. Using those resources allowed me to support square power-of-two textures, and nothing more. Everything else what works is a result of hard experiment-trial-and-error process, and I wouldn't like to continue it for a few NP2 textures.

**Comment by masagrator** [comment]
Pull request is about requesting change in code, right? Just I'm not a programmer, so I can't contribute to it.
I will write only what i found:
textures are rounded by 128 up. So if set size of this texture is 622x720, then real size is 640x768
That method is working flawlessly for me in rawtex for Switch textures.

**Comment by gildor2** [comment]
There's some rounding code for NSW decoder, but it rounds "blocks". And rounding value depends on texture size, I've found everything from experiments. I do not understand where else I should put "round to power of two" (or to what? 640 is not Power of 2).

**Comment by masagrator** [comment]
And it needs to be? Because textures are showing something different. 
Size of uexp for last file is 1921 kB which is exactly (640*768*4)/1024 + archive informations. 
Next file - 499x258. Size - 769 kB. (512*384*4)/1024=768
Biggest one from UI -  1880x1253. Size - 9600 kB (9 830 701 B). (1920*1280*4)/1024 = 9600 kB (9 830 400 B).
Everything looks correct.

**Comment by masagrator** [comment]
It looks like this method works only for b8g8r8a8. For DXT5 & DXT1 it looks slightly different, but still calculating resolution is vital in exported file.

Noesis for my DXT5 file 220x60 and it was broken. But when i putted constant to height "64" it untiled correctly.
So it looks like values needs to be calculated differently.

**Comment by gildor2** [comment]
Yes, it is nice when you can put constant by hands and immediately verify result. However I still can't see a uniform way for doing alignment. I've spent unreasonable time playing with Octopath textures, and what I see is that different values works for different cases.
